### ISSUE TITLE
The CUDA kernel of torch.nn.CrossEntropyLoss fails when the input tensor is too large

### ISSUE THREAD

**Issue by 1SAA** [pytorch/pytorch#85786]
### 🐛 Describe the bug

Here is the sample code `test.py`. Run it by `CUDA_LAUNCH_BLOCKING=1 python test.py`.

```python
import torch

n = 65536
m = 50257


a = torch.randn(n, m, device='cuda', requires_grad=True)
b = torch.randint(0, m, (n,), device='cuda')
c = torch.nn.CrossEntropyLoss()

loss = c(a, b)
print(loss)
loss.backward()
print("Passed")
```

It fails in forward when n is greater than 42730. The error information is below.

```
    return torch._C._nn.cross_entropy_loss(input, target, weight, _Reduction.get_enum(reduction), ignore_index, label_smoothing)
RuntimeError: CUDA error: an illegal memory access was encountered
```

When I changed m to 1024, it runs properly. So it is a problem occured when the input tensor is too large.
I also tested it in Pytorch1.11+cu113, it failed too.

## Environment
PyTorch version: 1.12.1+cu116
Is debug build: False
CUDA used to build PyTorch: 11.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.4 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.16.3
Libc version: glibc-2.31

Python version: 3.9.13 (main, Aug 25 2022, 23:26:10)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-48-generic-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 11.6.124
GPU models and configuration:
GPU 0: NVIDIA A100 80GB PCIe
GPU 1: NVIDIA A100 80GB PCIe
GPU 2: NVIDIA A100 80GB PCIe
GPU 3: NVIDIA A100 80GB PCIe
GPU 4: NVIDIA A100 80GB PCIe
GPU 5: NVIDIA A100 80GB PCIe
GPU 6: NVIDIA A100 80GB PCIe
GPU 7: NVIDIA A100 80GB PCIe

Nvidia driver version: 515.65.01
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.23.3
[pip3] torch==1.12.1+cu116
[pip3] torchaudio==0.12.1+cu116
[pip3] torchvision==0.13.1+cu116
[conda] numpy                     1.23.3                   pypi_0    pypi
[conda] torch                     1.12.1+cu116             pypi_0    pypi
[conda] torchaudio                0.12.1+cu116             pypi_0    pypi
[conda] torchvision               0.13.1+cu116             pypi_0    pypi

**Comment by t-vi** [comment]
For me the actual bound is 32769. I strongly suspect that you are running into a 32 bit indexing issue.

Indeed replacing int with int64_t in these two lines does get rid of the illegal memory access:
https://github.com/pytorch/pytorch/blob/614d6f19e3d30cac0d77059e738d1f25d75eb408/aten/src/ATen/native/cuda/Loss.cu#L236-L237

That said, the only thing it buys me is getting an out of memory error instead. I know that I'm running my tests on these measly 24GB cards, but so the options could be

- change to 64 bit indexing always and take a perf hit for everyone else using the kernel,
- add 64 bit indexing template and call 32 or 64 as applicable, adding many kernels,
- refuse to run on non 32-bit indexable tensors,
- hope that the code will be removed when we move to more codegen by default some day.

Depending on what we think about how often we run the native vs. nvfuser or some other, I'd probably favour the first or third/forth option.

**Comment by t-vi** [comment]
Indeed, thanks for pointing this out, @ptrblck . Given that #85395 has a fix, too, let's close this here and track things in #85005.
